src/delta.c

```diff
@@ -38,86 +38,159 @@
+		goto remove_and_fail;
 	}
-	do_delta(filename, file);
-	free_string(&filename);
+	if (!hash_equal(hash, to_hash)) {
+		fprintf(stderr, "Couldn't use delta file %s: application resulted in wrong hash\n", delta_file);
+		goto remove_and_fail;
+	}
+
+	goto end;
+
+remove_and_fail:
+	(void)remove(to_staged);
+end:
+	free_string(&prefixed_from_file);
 }
 
-static void do_delta(char *filename, struct file *file)
+/* Check if the delta filename is well-formed, if so return true and fill the from/to
+ * buffers with the corresponding hashes. Return false otherwise. */
+static bool check_delta_filename(const char *delta_name, char *from, char *to)
 {
-	char *origin;
-	char *dir, *base, *tmp = NULL, *tmp2 = NULL;
-	char *deltafile = NULL;
-	int ret;
-	struct stat sb;
+	/* Delta files have the form [FROM_VERSION]-[TO_VERSION]-[FROM_HASH]-[TO_HASH]. */
+	const char *s = delta_name;
+
+	/* Ignore versions, deltas will be used based on their hashes only. */
+	for (int i = 0; i < 2; i++) {
+		s = strchr(s, '-');
+		if (!s) {
+			return false;
+		}
+		s++;
+	}
 
-	string_or_die(&deltafile, "%s/delta/%i-%i-%s-%s", state_dir,
-		      file->deltapeer->last_change, file->last_change, file->deltapeer->hash, file->hash);
+	/* Note: SWUPD_HASH_LEN accounts for the NUL-terminator after the hash. */
+	const size_t hash_len = SWUPD_HASH_LEN - 1;
+	if (strlen(s) != (hash_len * 2 + 1)) {
+		return false;
+	}
 
-	ret = stat(deltafile, &sb);
-	if (ret != 0) {
-		free_string(&deltafile);
-		return;
+	if (s[hash_len] != '-') {
+		return false;
 	}
 
-	tmp = strdup(file->deltapeer->filename);
-	tmp2 = strdup(file->deltapeer->filename);
+	memcpy(from, s, hash_len);
 
-	dir = dirname(tmp);
-	base = basename(tmp2);
+	/* Consume the first hash and the separator. */
+	s += hash_len + 1;
 
-	string_or_die(&origin, "%s/%s/%s", path_prefix, dir, base);
+	memcpy(to, s, hash_len);
 
-	ret = apply_bsdiff_delta(origin, filename, deltafile);
-	if (ret) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
-	xattrs_copy(origin, filename);
+	return true;
+}
 
-	if (!verify_file(file, filename)) {
-		unlink_all_staged_content(file);
-		goto out;
-	}
+void apply_deltas(struct manifest *current_manifest)
+{
+	char *delta_dir;
+	string_or_die(&delta_dir, "%s/delta", state_dir);
 
-	if (xattrs_compare(origin, filename) != 0) {
-		unlink_all_staged_content(file);
-		goto out;
+	DIR *dir = opendir(delta_dir);
+	if (!dir) {
+		/* No deltas available to apply. */
+		return;
 	}
 
-	unlink(deltafile);
+	struct dirent *ent;
+	while ((ent = readdir(dir))) {
+		if (!strcmp(ent->d_name, ".") || !strcmp(ent->d_name, "..")) {
+			continue;
+		}
+
+		char *to_staged = NULL;
+
+		char *delta_name = ent->d_name;
+		char *delta_file;
+		string_or_die(&delta_file, "%s/%s", delta_dir, delta_name);
+
+		char from[SWUPD_HASH_LEN] = {};
+		char to[SWUPD_HASH_LEN] = {};
+		if (!check_delta_filename(delta_name, &from[0], &to[0])) {
+			fprintf(stderr, "Invalid name for delta file: %s\n", delta_file);
+			goto next;
```

Comment:
So this removes everything from delta, even if it doesn't look like a delta.

If we want to do multiple bundle processing in parallel then we need to have something like multiple delta directories, one per bundle.

Reply:
The original code was more conservative about deleting files, but previous reviews asked for more aggressive cleaning up.

As for multiple bundle processing, when we cross that bridge we see how it is the best fix. E.g. one can walk only through the deltas names in the bundle pack, instead of all files in the directory.